biz.aQute.bndlib/src/aQute/bnd/plugin/spi/SPIAnnotations.java

```diff
@@ -0,0 +1,248 @@
+
+		for (Entry<Annotation, Clazz> entry : entrySet) {
+			String fqcn = ((TypeRef) entry.getKey()
+				.get("value")).getFQN();
+			ArrayList<String> list = services.computeIfAbsent(fqcn, k -> new ArrayList<>());
+			list.add(entry.getValue()
+				.getClassName()
+				.getFQN());
+		}
+
+		for (Entry<String, ArrayList<String>> entry : services.entrySet()) {
+			String key = "META-INF/services/" + entry.getKey();
+			ArrayList<String> list = entry.getValue();
+			list.add(0, "# Generated by bnd");
+			String value = Strings.join("\n", entry.getValue());
+			a.getJar()
+				.putResource(key, new EmbeddedResource(value, a.lastModified()));
+		}
+	}
+
+	void append(Analyzer analyzer, String key, String value) {
+		analyzer.setProperty(key, Processor.append(analyzer.getProperty(key), value));
+	}
+
+	String toCapability(Entry<Annotation, Clazz> entry) {
+		TypeRef serviceType = entry.getKey()
+			.get("value");
+
+		StringBuilder sb = new StringBuilder();
+		sb.append("osgi.serviceloader;osgi.serviceloader='");
+		sb.append(serviceType.getFQN());
+		sb.append("'");
+		sb.append(";register:=");
+		sb.append(entry.getValue()
+			.getClassName()
+			.getFQN());
+
+		Annotation annotation = entry.getKey();
+
+		Object[] attributes = annotation.get("attribute");
+
+		if (attributes != null) {
+			for (Object attribute : attributes) {
+				sb.append(";");
+				sb.append(String.valueOf(attribute));
+			}
+		}
+
+		TypeRef[] uses = annotation.get("uses");
+
+		if (uses != null) {
+			sb.append(";uses:='");
+			sb.append(Arrays.stream(uses)
+				.map(tr -> tr.getPackageRef()
+					.getFQN())
+				.collect(Collectors.joining(",")));
+			sb.append("'");
+		}
+
+		return sb.toString();
+	}
+
+	String toRequirement(Entry<Annotation, Clazz> entry) {
+		try {
+			ServiceConsumer serviceConsumer = entry.getKey()
+				.getAnnotation();
+
+			TypeRef serviceType = entry.getKey()
+				.get("value");
+
+			StringBuilder sb = new StringBuilder();
+			sb.append("osgi.serviceloader;osgi.serviceloader='");
+			sb.append(serviceType.getFQN());
+			sb.append("';filter:='(osgi.serviceloader=");
+			sb.append(serviceType);
+			sb.append(")'");
+			sb.append(";cardinality:=");
+			sb.append(serviceConsumer.cardinality()
+				.toString()
+				.toLowerCase());
+			sb.append(";effective:=");
+			sb.append(serviceConsumer.effective()
+				.toString()
+				.toLowerCase());
+			sb.append(";resolution:=");
+			sb.append(serviceConsumer.resolution()
+				.toString()
+				.toLowerCase());
+
+			return sb.toString();
+		} catch (Exception e) {
+			throw Exceptions.duck(e);
+		}
+	}
+
+	private void processClazz(Analyzer analyzer, Clazz c, Map<Annotation, Clazz> consumerTypes,
+		Map<Annotation, Clazz> providerTypes) {
+
+		c.annotations(SERVICE_CONSUMERS)
+			.flatMap(a -> Arrays.stream((Object[]) a.get("value")))
+			.map(Annotation.class::cast)
+			.forEach(a -> consumerTypes.put(a, c));
+
+		c.annotations(SERVICE_CONSUMER)
+			.forEach(a -> consumerTypes.put(a, c));
+
+		c.methods()
```

Comment:
You can delete both calls to `Clazz.methods()` since we no longer allow the annotation on methods.

Reply:
good catch. updated!